src/flp/szk.rs

```diff
@@ -591,34 +610,56 @@ where
         ))
     }
 
-    /// Returns true if the verifier message indicates that the input from which
-    /// it was generated is valid.
+    pub(crate) fn merge_verifiers(
+        &self,
+        mut leader_share: SzkQueryShare<T::Field, SEED_SIZE>,
+        helper_share: SzkQueryShare<T::Field, SEED_SIZE>,
+    ) -> Result<SzkJointShare<SEED_SIZE>, SzkError> {
+        for (x, y) in leader_share
+            .flp_verifier
+            .iter_mut()
+            .zip(helper_share.flp_verifier)
+        {
+            *x += y;
+        }
+        if self.typ.decide(&leader_share.flp_verifier)? {
+            match (
+                leader_share.joint_rand_part_opt,
+                helper_share.joint_rand_part_opt,
+            ) {
+                (Some(leader_part), Some(helper_part)) => Ok(Some([leader_part, helper_part])),
+                (None, None) => Ok(None),
+                _ => Err(SzkError::InvalidState(
+                    "at least one of the joint randomness parts is missing".to_string(),
+                )),
+            }
+        } else {
+            Err(SzkError::Decide("failed to verify FLP proof".to_string()))
+        }
+    }
+    /// Returns true if the leader and helper derive identical joint randomness
+    /// seeds
     pub fn decide(
         &self,
-        verifier: SzkVerifier<T::Field, SEED_SIZE>,
         query_state: SzkQueryState<SEED_SIZE>,
-    ) -> Result<bool, SzkError> {
-        // Check if underlying FLP proof validates
-        let check_flp_proof = self.typ.decide(&verifier.flp_verifier)?;
-        if !check_flp_proof {
-            return Ok(false);
-        }
+        joint_share: &[Seed<SEED_SIZE>; 2],
```

Comment:
I think we could better encapsulate the SZK logic if this took in a `&SzkJointShare` instead. This would let us remove the match block from `Mastic::prepare_next()`.